machines: Implement reverting and deletion of VM snapshots

skobyda opened the pull request:
- [x] This PR builds on top of #13100 , so only last 3 commits are relevant to this PR.

Reverting to a snapshot means VM is restored to a state in which it was when a snapshot was taken.
![Screenshot from 2020-06-29 22-15-23](https://user-images.githubusercontent.com/42733240/86052106-bbdd0780-ba56-11ea-9502-e19a318cd87e.png)

Deleting snapshots is self explanatory.
![Screenshot from 2020-06-29 22-02-47](https://user-images.githubusercontent.com/42733240/86052118-c0a1bb80-ba56-11ea-8ef6-5f849a077ed1.png)


Also with action of reverting, it's important to distinguish which snapshot is "current". Current snapshot means that current state of VM is "closest", is succeeding "current" snapshot. So when you create several snapshots, let's say A, B, C, D in that order, at that moment D is current snapshot. But when you revert to the snapshot B, that snapshot becomes the current one.
I tried to solve this problem of distinguishing current snapshot by adding border on the right side of the row of the snapshot and changing background color to `--pf-global--BackgroundColor--200`. Although I don't know if this would be clear enough to the user.
![Screenshot from 2020-06-29 21-57-51](https://user-images.githubusercontent.com/42733240/86052127-c4cdd900-ba56-11ea-8746-55dd489d5d82.png)

garrett commented:
Is name mandatory? If so, I'd suggest it should be the first column, before date. Then I suppose it makes sense to have description follow it.

And then we could use a badge to indicate the active snapshot. I'd say use the "unread" style. It's a bummer as it's not semantic, but at least nothing about "unread" is presented in the UI, so it's not semantic anyway. This is probably a time we'd need aria, specifically the [aria-current](https://www.w3.org/TR/wai-aria-1.1/#aria-current) attribute. (The attribute `aria-current="true"` should be on the element, not the badge, as row itself is active, not the descendants in it.)

Some of this could go in #13100 (the reordering, for example).

Questions:

1. What happens when you revert? Do you lose changes to everything after that snapshot? Or can you bounce between snapshots?
2. How can you create snapshots?

andreasn commented:
> Is name mandatory? If so, I'd suggest it should be the first column, before date. Then I suppose it makes sense to have description follow it.

I feel strongly that Date should be the first column. Looking at how Boxes do it, their implementation is dates-only.
![Screenshot from 2020-06-30 11-06-26](https://user-images.githubusercontent.com/209895/86107146-c385c600-bac1-11ea-83ba-c9aabee7ae47.png)

Name is currently a generated string, but it's possible to change it to a custom string in the create dialog. My expectation is that most people will keep the auto-generated string.

andreasn commented on pkg/machines/components/vmSnapshotsRevertModal.jsx:
```diff
@@ -0,0 +1,91 @@
+                    <Modal.Title>{ cockpit.format(_("Revert to Snapshot $0"), snap.name) }</Modal.Title>
+                </Modal.Header>
+                <Modal.Body>
+                    { cockpit.format(_("Confirm this action")) }
```
Needs some language around how this is a possibly destructive action.

andreasn replied:
Same for the delete dialog. But I'm too shit at reading code, because I can't find that...

skobyda replied:
Yes. As I mentioned in my response to garrett's comment, we could put some warning saying something like you will loose all "non-snapshotted" progress when you revert, or/plus perhaps even allow user to create some sort of "temporary" snapshot before reverting so they do not loose all the progress.

andreasn replied:
Maybe something like: _"Reverting to this snapshot will take the VM back to the time of the snapshot and the current state will be lost, along with any data not snapshotted"_

Might be a bit long, so suggestions for tweaks welcome!

andreasn replied:
_plus perhaps even allow user to create some sort of "temporary" snapshot before reverting so they do not loose all the progress._

Not a bad idea, but let's do that in another PR in that case.

skobyda replied:
> Maybe something like: _"Reverting to this snapshot will take the VM back to the time of the snapshot and the current state will be lost, along with any data not snapshotted"_

"snapshotted" looks like a weird word, perhaps "captured in a snapshot" ?

Basically by loosing state and data we only mean loosing data on disk and in memory (that's the only thing snapshots capture). And in some snapshot variations (some are only to be implemented by libvirt), some snapshots only capture VM's memory, and some only capture VM's disks state.. Changes to external things like Network interfaces or changing number of CPUs are not captured in snapshots. 

So I cannot thing of sentence which says that disk data might be lost ONLY if they are captured and thus overwritten in the snapshot, same with memory and keep it short

andreasn commented on pkg/machines/components/vmSnapshotsTab.css:
```diff
@@ -0,0 +1,19 @@
+
+.snap-current {
+    border-left: 3px solid var(--pf-global--primary-color--100);
+    background-color: var(--pf-global--BackgroundColor--200)
```
I was looking around at what other projects do for this.
* Virt-manager: checkmark (✔️)
* GNOME Boxes: square is different color [◼️]
* OpenShift: [Nothing particular](http://openshift.github.io/openshift-origin-design/designs/virtualization/4.3/snapshots/)
* oVirt: [Change date and description](https://3.bp.blogspot.com/-U8QMcx9JFA4/XBhjV5hFI1I/AAAAAAAAAgw/eAPNOt_hWh087A9Ns2vrRsYWnUcitpkUgCLcBGAs/s1600/Selection_999%2528836%2529.png)

I kind of like the oVirt approach, but just a small visual indicator like this works too.

garrett replied:
The selected visual state from PF4 is for _interactive_ selections (like nav, tab, or a selected row in a table), not a way to show something is active or current. And it needs to be obvious not just to screen readers (with ARIA) but also visually in the browser. The blue line "doesn't cut it" (is not good enough), in other words. 

The others aren't really obvious either (from your description, at least). It's nice to have a visual indicator, but if it is relying only on color or is just an undefined glyph, it isn't enough. (That could be changed with a tooltip for the icon and appropriate ARIA on the active element.)

skobyda replied:
> (That could be changed with a tooltip for the icon and appropriate ARIA on the active element.)

What icon could be best for highlighting active/current snapshot?

garrett replied:
PF4 happens to use `fa-check-circle` in a label their example for "running" on the icons page:

![status icons](https://www.patternfly.org/v4/img/ffc62ca48f6c8d914f97acff46e8e893-sm-icons-1.png)
(_These look large here, since I guess it's a 2× screenshot. They should not look so large when used in Cockpit, of course._)

- Icons page: https://www.patternfly.org/v4/design-guidelines/styles/icons
- Label component: https://www.patternfly.org/v4/documentation/react/components/label

This mirrors my suggestion of using an "active" badge (I really meant label, but labels are newer to PF and are extremely similar to badges anyway).

If space is of concern, we could just use the icon alone with a tooltip (and use ARIA on the selected item, as I suggested above).

If we're using a label, it should go _after_ the text; if an icon, it should go into its own column preceding everything (with the others only having empty space where the icon would be).

FWIW: Regardless of implementation, the date and description could also change.

andreasn replied:
So something like this?
![Screenshot from 2020-07-01 12-57-16](https://user-images.githubusercontent.com/209895/86236605-a7a22300-bb9a-11ea-8d37-120b1df14d9c.png)

garrett replied:
Sure, that works for me! Thanks! (Once it's "shut off".)

andreasn replied:
Or maybe like this?

![Screenshot from 2020-07-01 12-51-13](https://user-images.githubusercontent.com/209895/86237020-4d559200-bb9b-11ea-9f63-57af6755d7b4.png)

skobyda replied:
I like first mockup. Getting rid of Time fully and replacing it with icon is not nice, I think user should be able to see time of creation at all time (even if they decide to not include that in snapshotname)..

andreasn replied:
Sure, let's go for the first one!

skobyda commented:
> Is name mandatory? If so, I'd suggest it should be the first column, before date. Then I suppose it makes sense to have description follow it.

Libvirt doesn't demand it when creating snapshots, but then it puts unix timestamp as snapshot name, which from then on is identifying property of snapshot, which I do not find very user friendly to have unixtimestamps as names. So instead in the creation of snapshots I plan to pregenerate the name with VM name + date and time. Of course, user is able to change that if they want to. (more in https://github.com/cockpit-project/cockpit/pull/14259)

> And then we could use a badge to indicate the active snapshot. I'd say use the "unread" style. It's a bummer as it's not semantic, but at least nothing about "unread" is presented in the UI, so it's not semantic anyway. This is probably a time we'd need aria, specifically the [aria-current](https://www.w3.org/TR/wai-aria-1.1/#aria-current) attribute. (The attribute `aria-current="true"` should be on the element, not the badge, as row itself is active, not the descendants in it.)
> 
> Some of this could go in #13100 (the reordering, for example).
> 
> Questions:
> 
> 1. What happens when you revert? Do you lose changes to everything after that snapshot? Or can you bounce between snapshots?

Yes you loose them. If you do not want to loose them, you have to create a snapshot before reverting to another snapshot. (Perhaps we could put warning of some sort, or have option in revert to snapshot dialog in a sense of "create snapshot if current state and then revert")

> 2. How can you create snapshots?

There is a PR for that https://github.com/cockpit-project/cockpit/pull/14259

skobyda commented:
This is weird. I tried to User PF4 Label with Icon, just like here https://www.patternfly.org/v4/documentation/react/components/label#outline, and instead it renders blue label with no icon. anybody encountered something like this before? Perhaps PF4 bug?
![Screenshot from 2020-07-02 17-01-43](https://user-images.githubusercontent.com/42733240/86375384-e2cd5080-bc85-11ea-86dd-abd7e886f966.png)

garrett commented:
> Yes you loose them. If you do not want to loose them, you have to create a snapshot before reverting to another snapshot. (Perhaps we could put warning of some sort, or have option in revert to snapshot dialog in a sense of "create snapshot of current state and then revert")


Oh, we should make it obvious that those go away. Perhaps something to the effect of:

---

## Revert to SNAPSHOT_NAME_HERE

Reverting to **SNAPSHOT_NAME_HERE** will discard current changes to the current virtual machine and remove all snapshots made afterward, including:

- Snapshot1 (2020-05-22)
- Snapshot2 (2020-05-22)

[ Revert ]  Cancel 

---

And we'd have different text if it doesn't remove anything. Perhaps something like:

Reverting to **SNAPSHOT_NAME_HERE** will discard current changes to the virtual machine.

(And the title and action buttons would be the same.)

garrett commented:
> This is weird. I tried to User PF4 Label with Icon, just like here https://www.patternfly.org/v4/documentation/react/components/label#outline, and instead it renders blue label with no icon. anybody encountered something like this before? Perhaps PF4 bug?
> ![Screenshot from 2020-07-02 17-01-43](https://user-images.githubusercontent.com/42733240/86375384-e2cd5080-bc85-11ea-86dd-abd7e886f966.png)

This might also be my fault due to some code in the lists that makes assumptions. They're spans with pf-c-label in the HTML implementation. I don't see anything listed as pf-c-label, so I don't think this is the right answer. (It's my first guess and might lead to a solution though, so I'm leaving it here.)

Are you using the latest version of PatternFly? I think the labels are brand new. Old versions won't do the right thing.

martinpitt commented:
#13100 landed; so this can be rebased and finished.

skobyda commented:
@garrett So I updated the PF library, done few changes in CSS, and here is how it looks now:

![Screenshot from 2020-08-03 12-17-28](https://user-images.githubusercontent.com/42733240/89172886-a9894880-d583-11ea-94d7-e57a3c26b0da.png)
Do you know what is the proper approach to make the Current icon be centered with the text to its left?
Also, there are problems with columns here, but they are not caused by this PR. I reported it here: https://github.com/cockpit-project/cockpit/issues/14443

andreasn commented:
How can two different snapshots both be the current snapshot?

![Screenshot from 2020-08-11 13-49-00](https://user-images.githubusercontent.com/209895/89893866-7117e900-dbd9-11ea-82be-4550a935fb16.png)

andreasn commented:
![Screenshot from 2020-08-11 14-28-00](https://user-images.githubusercontent.com/209895/89897141-e1753900-dbde-11ea-8fd5-31842e633d95.png)

This text feels a little bare still.

garrett commented:
There are still some issues with the list. Is this rebased to master? (I don't remember which PR I included a fix for stuff like that. Looking now...)

garrett commented:
Drat. I think the fix is @ https://github.com/cockpit-project/cockpit/pull/14346/commits/a2f29ed85a1a0b2563d0aa4ef04dec5cf2b6d8c2 which is in https://github.com/cockpit-project/cockpit/pull/14346 and depends on the detail sub-pages.

The alignment is way off other than just that fix though. I'll try yet again to get this working locally again.

KKoukiou commented on pkg/lib/cockpit-components-listing-panel.jsx:
```diff
@@ -26,6 +26,7 @@ import './cockpit-components-listing-panel.scss';
  *     - name tab name (has to be unique in the entry, used as react key)
  *     - renderer react component
  *     - data render data passed to the tab renderer
+ *     - hide flag whetever tab link and it's content should be shown
```
This sentence has some grammar issues. Why do you need the hide flag? What if you just don't pass it?

skobyda replied:
Fixed

KKoukiou reviewed (changes requested):
missing Closes#, missing tests. codewise looks good.

skobyda commented:
I tried to work a bit with spacing, but it still looks weird. Using pficon-ok should be the same as pficon-pending, since I added it to some scss files, but it still destroys spacing. Any idea what might go wrong?
![Screenshot from 2020-09-01 14-43-37](https://user-images.githubusercontent.com/42733240/91853127-0b5fcf80-ec62-11ea-9da7-313764bb04e7.png)

andreasn commented on pkg/machines/components/vmSnapshotsTab.jsx:
```diff
@@ -146,6 +153,49 @@ class VmSnapshotsTab extends React.Component {
+                    };
+
+                    return (
+                        <div className='machines-listing-actions pull-right'>
```
This seems to be why it breaks!
Remove `pull-right`, and you should be good.

andreasn replied:
This is how it looks without that class applied.
![Screenshot from 2020-09-01 16-54-02](https://user-images.githubusercontent.com/209895/91867297-c6dd2f80-ec73-11ea-907d-22d9c4ec44a6.png)

skobyda replied:
Thanks a lot for your help. I fixed it and this is how it looks now (basically the same as in your screenshot): 
![Screenshot from 2020-09-02 02-01-05](https://user-images.githubusercontent.com/42733240/91917486-590b2500-ecc0-11ea-81c6-986a5d13bcc6.png)

skobyda commented:
> ![Screenshot from 2020-08-11 14-28-00](https://user-images.githubusercontent.com/209895/89897141-e1753900-dbde-11ea-8fd5-31842e633d95.png)
> 
> This text feels a little bare still.

So I rephrased the text a bit to be more specific for this case
![Screenshot from 2020-09-02 02-01-16](https://user-images.githubusercontent.com/42733240/91917453-44c72800-ecc0-11ea-9eea-dfa9c35af475.png)

skobyda commented:
@KKoukiou I added the test, updated the PR. Could you give it a code review?
also @andreasn @garrett, could one of you give a glance if you approve the current design?

andreasn commented on pkg/machines/components/vmSnapshotsRevertModal.jsx:
```diff
@@ -0,0 +1,91 @@
+                    <Modal.Title>{ cockpit.format(_("Revert to Snapshot $0"), snap.name) }</Modal.Title>
+                </Modal.Header>
+                <Modal.Body>
+                    { cockpit.format(_("Reverting to this snapshot will take the VM back to the time of the snapshot and the current state will be lost, along with any data not caputred in a snapshot")) }
```
> caputred

Typo. Should be _Captured_

skobyda replied:
Fixed that

andreasn reviewed (approved):
Apart from the small typo, looks good to me!

KKoukiou commented on test/verify/check-machines:
```diff
@@ -3467,6 +3467,42 @@ class TestMachines(MachineCase, StorageHelpers, NetworkHelpers):
+        m = self.machine
+
+        if m.image in distrosWithoutSnapshot:
+            return
```
We should have the negative check as well, in this case that the snapshots tab is missing. With this approach we are notified when an image refresh happends, and we get the required version of libvirt-dbus and we can update the list.

Otherwise, we will probably just forget the list there and not test it.

skobyda replied:
Fixed. I added a check for the snapshots tab not being present. So once some distro will get a new version of libvirt-dbus, it will fail and I will be forced to update the distrosWithoutSnapshot list.
```
        if m.image in distrosWithoutSnapshot:
            b.wait_not_present("#vm-subVmTest1-snapshots")
            return
```

KKoukiou reviewed (dismissed):
Thanks for the fixes, please merge once tests get green.

skobyda commented:
No machines test failure
